pycaw/callbacks.py

```diff
@@ -1,3 +1,108 @@
-# TODO
-# implement easy to use callbacks
-# (is already written, but 1 pull at the time)
+from ctypes import pointer
+
+from comtypes import COMObject
+
+from pycaw.api.audiopolicy import (IAudioSessionControl2, IAudioSessionEvents,
+                                   IAudioSessionNotification)
+from pycaw.api.endpointvolume import IAudioEndpointVolumeCallback
+from pycaw.utils import AudioSession
+
+
+class AudioSessionNotification(COMObject):
+    """
+    SESSION CREATED CALLBACK
+
+    """
+    _com_interfaces_ = [IAudioSessionNotification]
+
+    # "OnSessionCreated" is fired by Windows everytime a new session is created
+    def OnSessionCreated(self, NewSession):
+        ctl2 = NewSession.QueryInterface(IAudioSessionControl2)
+        new_session = AudioSession(ctl2)
+
+        self.on_session_created(new_session)
+
+    # pycaw user interface:
+    def on_session_created(self, new_session):
+        pass
+
+
+class AudioSessionEvents(COMObject):
+    """
+    AUDIO SESSION CALLBACK
+    """
+    _com_interfaces_ = [IAudioSessionEvents]
+
+    # ======= DECODE RETURNED INT VALUE =======
+    # see audiosessiontypes.h and audiopolicy.h
+    AudioSessionState = [
+        "Inactive",
+        "Active",
+        "Expired"
+    ]
+
+    AudioSessionDisconnectReason = [
+        "DeviceRemoval",
+        "ServerShutdown",
+        "FormatChanged",
+        "SessionLogoff",
+        "SessionDisconnected",
+        "ExclusiveModeOverride"
+    ]
+
+    def OnSimpleVolumeChanged(self, NewVolume, NewMute, EventContext):
+        self.on_simple_volume_changed(NewVolume, NewMute, EventContext)
+
+    def OnStateChanged(self, NewState):
+        new_state = self.AudioSessionState[NewState]
+        new_state_id = NewState
+        self.on_state_changed(new_state, new_state_id)
+
+    def OnSessionDisconnected(self, DisconnectReason):
+        disconnect_reason = self.AudioSessionDisconnectReason[DisconnectReason]
+        disconnect_reason_id = DisconnectReason
+        self.on_session_disconnected(disconnect_reason, disconnect_reason_id)
+
+    # pycaw user interfaces:
```

Comment:
https://docs.microsoft.com/en-us/windows/win32/api/audiopolicy/nn-audiopolicy-iaudiosessionevents

The original API for IAudioSessionEvents is much longer (7) I only implemented 3 com methods. Do you think it would get tedious to force the implementation of methods that may later get added like ::OnIconPathChanged?

Perhaps we can force implement only the current (as they are the most important too).
And in the future the other 4 will get a pass.